diff -rNU3 a/bfd/config.bfd b/bfd/config.bfd
--- a/bfd/config.bfd	2023-02-23 14:44:07.778654904 +0900
+++ b/bfd/config.bfd	2023-02-23 13:03:01.515839000 +0900
@@ -411,6 +411,7 @@
     targ_selvecs=arm_elf32_be_vec
     ;;
   arm-*-elf | arm*-*-freebsd* | arm*-*-linux-* | arm*-*-conix* | \
+  arm*-*-minix* | \
   arm*-*-uclinux* | arm-*-kfreebsd*-gnu | \
   arm*-*-eabi* | arm-*-rtems* | arm*-*-uclinuxfdpiceabi)
     targ_defvec=arm_elf32_le_vec
@@ -636,6 +637,10 @@
 	targ_cflags=-DOLD_FREEBSD_ABI_LABEL ;;
     esac
     ;;
+  i[3-7]86-*-minix)
+    targ_defvec=i386_elf32_minix_vec
+    targ_selvecs="iamcu_elf32_vec i386_coff_vec i386_pei_vec"
+    ;;
   i[3-7]86-*-netbsd* | i[3-7]86-*-knetbsd*-gnu)
     targ_defvec=i386_elf32_vec
     targ_selvecs="iamcu_elf32_vec i386_coff_vec i386_pei_vec"
diff -rNU3 a/bfd/configure b/bfd/configure
--- a/bfd/configure	2023-02-23 14:44:07.782651406 +0900
+++ b/bfd/configure	2023-02-23 13:03:01.519840000 +0900
@@ -13429,6 +13429,7 @@
     i386_coff_lynx_vec)		 tb="$tb cf-i386lynx.lo lynx-core.lo $coff" ;;
     i386_elf32_vec)		 tb="$tb elf32-i386.lo $elfxx_x86 elf32.lo $elf" ;;
     i386_elf32_fbsd_vec)	 tb="$tb elf32-i386.lo $elfxx_x86 elf32.lo $elf" ;;
+    i386_elf32_minix_vec)	 tb="$tb elf32-i386.lo $elfxx_x86 elf32.lo $elf" ;;
     i386_elf32_sol2_vec)	 tb="$tb elf32-i386.lo $elfxx_x86 elf32.lo $elf" ;;
     i386_elf32_vxworks_vec)	 tb="$tb elf32-i386.lo $elfxx_x86 elf32.lo $elf" ;;
     i386_mach_o_vec)		 tb="$tb mach-o-i386.lo" ;;
diff -rNU3 a/bfd/elf32-i386.c b/bfd/elf32-i386.c
--- a/bfd/elf32-i386.c	2023-02-23 14:44:07.794640913 +0900
+++ b/bfd/elf32-i386.c	2023-02-23 13:19:51.669122000 +0900
@@ -4465,6 +4465,16 @@
 
 #define	elf32_bed			      elf32_i386_bed
 
+/* MINIX3 support.  */
+
+#undef	TARGET_LITTLE_SYM
+#define	TARGET_LITTLE_SYM		i386_elf32_minix_vec
+#undef	TARGET_LITTLE_NAME
+#define	TARGET_LITTLE_NAME		"elf32-i386-minix"
+
+#undef	elf32_bed
+#define	elf32_bed				elf32_i386_minix_bed
+
 #include "elf32-target.h"
 
 /* FreeBSD support.  */
diff -rNU3 a/bfd/targets.c b/bfd/targets.c
--- a/bfd/targets.c	2023-02-23 14:44:07.858584952 +0900
+++ b/bfd/targets.c	2023-02-23 13:21:43.815258000 +0900
@@ -748,6 +748,7 @@
 extern const bfd_target i386_coff_lynx_vec;
 extern const bfd_target i386_elf32_vec;
 extern const bfd_target i386_elf32_fbsd_vec;
+extern const bfd_target i386_elf32_minix_vec;
 extern const bfd_target i386_elf32_sol2_vec;
 extern const bfd_target i386_elf32_vxworks_vec;
 extern const bfd_target i386_mach_o_vec;
diff -rNU3 a/bfd/version.h b/bfd/version.h
--- a/bfd/version.h	2023-02-23 14:44:07.862581454 +0900
+++ b/bfd/version.h	2022-08-05 16:38:29.000000000 +0900
@@ -18,5 +18,5 @@
    sonames.  */
 #define BFD_VERSION_DATE 20220805
 #define BFD_VERSION @bfd_version@
-#define BFD_VERSION_STRING  @bfd_version_package@ " " @bfd_version_string@
+#define BFD_VERSION_STRING  @bfd_version_package@ @bfd_version_string@
 #define REPORT_BUGS_TO @report_bugs_to@
diff -rNU3 a/config.guess b/config.guess
--- a/config.guess	2022-07-08 18:46:47.000000000 +0900
+++ b/config.guess	2023-02-23 13:47:36.699107000 +0900
@@ -967,7 +967,7 @@
 	GUESS=$UNAME_MACHINE-unknown-$GNU_SYS$GNU_REL-$LIBC
 	;;
     *:Minix:*:*)
-	GUESS=$UNAME_MACHINE-unknown-minix
+	GUESS=$UNAME_MACHINE-elf32-minix
 	;;
     aarch64:Linux:*:*)
 	GUESS=$UNAME_MACHINE-unknown-linux-$LIBC
diff -rNU3 a/configure b/configure
--- a/configure	2022-07-08 18:55:40.000000000 +0900
+++ b/configure	2023-02-23 13:42:35.160460000 +0900
@@ -3016,7 +3016,7 @@
       *-*-elf* | *-*-sysv4* | *-*-unixware* | *-*-eabi* | hppa*64*-*-hpux* \
       | *-*-linux* | *-*-gnu* | frv-*-uclinux* | *-*-irix5* | *-*-irix6* \
       | *-*-netbsd* | *-*-openbsd* | *-*-freebsd* | *-*-dragonfly* \
-      | *-*-solaris2* | *-*-nto* | *-*-nacl* | *-*-haiku*)
+      | *-*-solaris2* | *-*-nto* | *-*-nacl* | *-*-haiku* | *-*-minix*)
         case "${target}" in
           *-*-linux*aout* | *-*-linux*oldld*)
             ;;
diff -rNU3 a/gas/config/tc-i386.c b/gas/config/tc-i386.c
--- a/gas/config/tc-i386.c	2022-07-08 18:46:47.000000000 +0900
+++ b/gas/config/tc-i386.c	2023-02-23 13:36:24.946443000 +0900
@@ -474,6 +474,7 @@
      && !defined (TE_Haiku)				\
      && !defined (TE_FreeBSD)				\
      && !defined (TE_DragonFly)				\
+	 && !defined (TE_MINIX)				\
      && !defined (TE_NetBSD))
 /* This array holds the chars that always start a comment.  If the
    pre-processor is disabled, these aren't very useful.  The option
diff -rNU3 a/gas/config/tc-i386.h b/gas/config/tc-i386.h
--- a/gas/config/tc-i386.h	2022-07-08 18:46:47.000000000 +0900
+++ b/gas/config/tc-i386.h	2023-02-23 13:38:07.423888000 +0900
@@ -68,6 +68,10 @@
 #define ELF_TARGET_FORMAT64	"elf64-x86-64-sol2"
 #endif
 
+#ifdef TE_MINIX
+#define ELF_TARGET_FORMAT	"elf32-i386-minix"
+#endif
+
 #ifndef ELF_TARGET_FORMAT
 #define ELF_TARGET_FORMAT	"elf32-i386"
 #endif
diff -rNU3 a/gas/config/te-minix.h b/gas/config/te-minix.h
--- a/gas/config/te-minix.h	1970-01-01 09:00:00.000000000 +0900
+++ b/gas/config/te-minix.h	2021-04-05 01:29:44.000000000 +0900
@@ -0,0 +1,9 @@
+#define TE_MINIX	1
+
+/* Added these, because if we don't know what we're targeting we may
+   need an assembler version of libgcc, and that will use local
+   labels.  */
+#define LOCAL_LABELS_DOLLAR 1
+#define LOCAL_LABELS_FB 1
+
+#include "obj-format.h"
diff -rNU3 a/gas/configure b/gas/configure
--- a/gas/configure	2023-02-23 14:44:07.906542980 +0900
+++ b/gas/configure	2023-02-23 13:31:46.954262000 +0900
@@ -5607,6 +5607,10 @@
   fi
   ;;
 
+minix*)
+  lt_cv_deplibs_check_method=pass_all
+  ;;
+
 gnu*)
   lt_cv_deplibs_check_method=pass_all
   ;;
diff -rNU3 a/gas/configure.tgt b/gas/configure.tgt
--- a/gas/configure.tgt	2023-02-23 14:44:07.906542980 +0900
+++ b/gas/configure.tgt	2023-02-23 13:27:20.837474000 +0900
@@ -443,6 +443,8 @@
   *-*-elf | *-*-rtems* | *-*-sysv4*)	fmt=elf ;;
   *-*-solaris*)				fmt=elf em=solaris ;;
   *-*-aros*)				fmt=elf em=linux ;;
+  i*-*-minix*)				fmt=elf em=minix ;;
+  arm*-*-minix*)			fmt=elf em=armeabi ;;
   *-*-vxworks* | *-*-windiss)		fmt=elf em=vxworks ;;
 esac
 

diff -rNU3 a/ld/configure.tgt b/ld/configure.tgt
--- a/ld/configure.tgt	2023-02-23 14:44:07.930521996 +0900
+++ b/ld/configure.tgt	2023-02-23 13:03:01.519840000 +0900
@@ -209,6 +209,9 @@
 arm-*-elf | arm*-*-eabi* | arm-*-rtems* | arm-*-genode*)
 			targ_emul=armelf
 			;;
+arm*-*-minix*)		targ_emul=armelf_minix
+			targ_extra_emuls="armelf"
+			;;
 arm-*-kaos*)		targ_emul=armelf
 			;;
 arm9e-*-elf)		targ_emul=armelf
@@ -384,6 +387,9 @@
 			targ64_extra_emuls="elf_x86_64 elf32_x86_64"
 			targ64_extra_libpath="elf_x86_64 elf32_x86_64"
 			;;
+i[3-7]86-*-minix)	targ_emul=elf_i386_minix
+			targ_extra_emuls="elf_i386"
+			;;
 i[3-7]86-*-redox*)	targ_emul=elf_i386
 			targ_extra_emuls=elf_x86_64
 			;;
diff -rNU3 a/ld/emulparams/armelf_minix.sh b/ld/emulparams/armelf_minix.sh
--- a/ld/emulparams/armelf_minix.sh	1970-01-01 09:00:00.000000000 +0900
+++ b/ld/emulparams/armelf_minix.sh	2023-02-23 13:03:01.519840000 +0900
@@ -0,0 +1,15 @@
+. ${srcdir}/emulparams/armelf.sh
+. ${srcdir}/emulparams/elf_minix.sh
+OUTPUT_FORMAT="elf32-littlearm"
+MAXPAGESIZE="CONSTANT (MAXPAGESIZE)"
+COMMONPAGESIZE="CONSTANT (COMMONPAGESIZE)"
+
+DATA_START_SYMBOLS='PROVIDE (__data_start = .);';
+
+# Dynamic libraries support
+GENERATE_SHLIB_SCRIPT=yes
+TARGET2_TYPE=got-rel
+
+GENERATE_PIE_SCRIPT=yes
+
+unset EMBEDDED
diff -rNU3 a/ld/emulparams/elf_i386_minix.sh b/ld/emulparams/elf_i386_minix.sh
--- a/ld/emulparams/elf_i386_minix.sh	1970-01-01 09:00:00.000000000 +0900
+++ b/ld/emulparams/elf_i386_minix.sh	2023-02-23 13:03:01.519840000 +0900
@@ -0,0 +1,3 @@
+. ${srcdir}/emulparams/elf_i386.sh
+. ${srcdir}/emulparams/elf_minix.sh
+OUTPUT_FORMAT="elf32-i386-minix"
diff -rNU3 a/ld/emulparams/elf_minix.sh b/ld/emulparams/elf_minix.sh
--- a/ld/emulparams/elf_minix.sh	1970-01-01 09:00:00.000000000 +0900
+++ b/ld/emulparams/elf_minix.sh	2023-02-23 13:03:01.523840000 +0900
@@ -0,0 +1 @@
+ELF_INTERPRETER_NAME=\"/libexec/ld-elf.so.1\"
diff -rNU3 a/ld/Makefile.in b/ld/Makefile.in
--- a/ld/Makefile.in	2023-02-23 14:44:07.950504506 +0900
+++ b/ld/Makefile.in	2023-02-23 13:03:01.523840000 +0900
@@ -668,6 +668,7 @@
 	earmelf_linux.c \
 	earmelf_linux_eabi.c \
 	earmelf_linux_fdpiceabi.c \
+	earmelf_minix.c \
 	earmelf_nacl.c \
 	earmelf_nbsd.c \
 	earmelf_nbsd_eabi.c \
@@ -789,6 +790,7 @@
 	eelf_i386_fbsd.c \
 	eelf_i386_haiku.c \
 	eelf_i386_ldso.c \
+	eelf_i386_minix.c \
 	eelf_i386_sol2.c \
 	eelf_i386_vxworks.c \
 	eelf_iamcu.c \
@@ -1470,6 +1472,7 @@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf_i386_fbsd.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf_i386_haiku.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf_i386_ldso.Po@am__quote@
+@AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf_i386_minix.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf_i386_sol2.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf_i386_vxworks.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf_iamcu.Po@am__quote@
@@ -2355,6 +2358,7 @@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/earmelf_linux.Pc@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/earmelf_linux_eabi.Pc@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/earmelf_linux_fdpiceabi.Pc@am__quote@
+@AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/earmelf_minix.Pc@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/earmelf_nacl.Pc@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/earmelf_nbsd.Pc@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/earmelf_phoenix.Pc@am__quote@
